[AWS] Awscall to F[_] Transformation #730

yisraelU opened the pull request:
closes #730
Currently throwing an exception as F has no structure.
should I add a context bound ?

yisraelU committed:
aws interpreter transformation

yisraelU committed:
aws interpreter transformation

yisraelU committed:
add AwsOperationKind.Unary.unary

kubukoz commented:
I'm honestly surprised this compiles on all Scala versions

kubukoz commented:
> Currently throwing an exception as F has no structure.
> should I add a context bound ?

I think that'd be a safer approach - if someone calls this transformation on a service with streaming- and non-streaming operations, we won't break the whole thing if they're only going to call some of these methods...

kubukoz commented on modules/aws/src/smithy4s/aws/package.scala:
```diff
@@ -32,4 +33,22 @@ package object aws {
     MonadThrow[F].catchNonFatal(new String(bytes, "UTF-8"))
 
+  // format: off
+  def simplify[Alg[_[_, _, _, _, _]], F[_]](service: Service[Alg]): service.Interpreter[AwsCall[F, *, *, *, *, *]] => service.FunctorInterpreter[F] = {
```
I'm wondering if it'd be more convenient for the callers if `service` was implicit and `interpreter` was a normal function parameter.

yisraelU commented:
> I'm honestly surprised this compiles on all Scala versions

Are you referring to the implicit ev requirement?

yisraelU commented:
> > Currently throwing an exception as F has no structure.
> > should I add a context bound ?
> 
> I think that'd be a safer approach - if someone calls this transformation on a service with streaming- and non-streaming operations, we won't break the whole thing if they're only going to call some of these methods...

Yeah of course.
I just wasn't sure what @Baccata had in mind as he provided the signature

kubukoz commented:
> Are you referring to the implicit ev requirement?

No, referring to how the streaming types are unified after the pattern match so that you can call `.run` without extra trickery. I've seen plenty of GADT issues back in the day, so it's good to see it working right

yisraelU committed:
require ApplicativeThrow instance

yisraelU commented:
> > Are you referring to the implicit ev requirement?
> 
> No, referring to how the streaming types are unified after the pattern match so that you can call `.run` without extra trickery. I've seen plenty of GADT issues back in the day, so it's good to see it working right

yeah thats what I meant by the evidence requirement , the fact that it recognizes the proof of the streaming types via the pattern match

Baccata commented:
Cool, I too wasn't sure it'd compile on all Scala versions. 

> should I add a context bound ? 

Yup, good idea, it eluded me when I wrote the sig. 

Alright, let's: 

1. ~add a `MonadThrow` constraint in there so that the exception is raised in `F`~ already done, gj !
2. change the exception type to `IllegalAccessException`
3. make that simplify method package-private 
4. add a `buildSimple` method new to [this one](https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/aws/src/smithy4s/aws/AwsClient.scala#L81-L83) so that users can have a functor client easily

Baccata commented on modules/aws/src/smithy4s/aws/package.scala:
```diff
@@ -32,4 +33,24 @@ package object aws {
+  def simplify[Alg[_[_, _, _, _, _]], F[_]:ApplicativeThrow](service: Service[Alg]): service.Interpreter[AwsCall[F, *, *, *, *, *]] => service.FunctorInterpreter[F] = {
+    interpreter =>
+     new PolyFunction5[service.Operation, Kind1[F]#toKind5] {
+      override def apply[A0, A1, A2, A3, A4](op: service.Operation[A0, A1, A2, A3, A4]): F[A2] = {
```
let's call those `I, E, O, SI, SO` as usual

yisraelU replied:
yeah I should have though of that

Baccata commented on modules/aws/src/smithy4s/aws/package.scala:
```diff
@@ -16,8 +16,9 @@
-
+import cats.{ApplicativeThrow, MonadThrow}
+import smithy4s.kinds._
+import smithy4s.aws.AwsOperationKind.Unary.unary
```
let's add a blank line after this (weird that scalafmt wouldn't say anything)

yisraelU commented:
@Baccata I thought ApplicativeThrow would be better for this function as principle of least power

yisraelU committed:
some cleanup plus
simplify handles error case with the exception type to IllegalAccessException
simplify method is package-private
add a buildSimple method new to AwsClient so that users can have a functor client easily

yisraelU committed:
this was not detected by scalafmt, I wonder why

Baccata reviewed (approved):
awesome !
